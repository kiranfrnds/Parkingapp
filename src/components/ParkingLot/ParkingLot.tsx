import {
  Button,
  Card,
  CardContent,
  TextField,
  Typography,
} from "@mui/material";
import { Container } from "@mui/system";
import React, { useContext, useState } from "react";
import { useNavigate } from "react-router-dom";
import { ParkingContext } from "../../ParkingContext/ParkingContext";

const ParkingLot = () => {
  const [carNumber, setCarNumber] = useState<any>("");
  const { parkingSlots, setParkingSlots, setStateId } =
    useContext(ParkingContext);
  const navigate = useNavigate();

  const handleCarNumber = (event: React.ChangeEvent<HTMLInputElement>) => {
    setCarNumber(event.target.value);
  };

  const handleParking = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.preventDefault();
    let temp = [...parkingSlots];
    let freeParkingSlots = temp.filter((item: any) => item.available === true);

    if (freeParkingSlots.length === 0) {
      alert("Parking Slots are Full");
    } else {
      const id = freeParkingSlots[Math.floor(Math.random() * temp.length)].id;
      const index = temp.findIndex((i) => i.id === id);
      temp[index].carNumber = carNumber;
      temp[index].available = false;
      temp[index].startTime = new Date();
      setParkingSlots(temp);
      setCarNumber("");
    }
  };

  const handleExit = (id: any) => {
    setStateId(id);
    navigate("/exit");
  };

  return (
    <>
      <Container
        sx={{
          width: 600,
          height: 200,
          display: "flex",
          justifyContent: "center",
          alignItems: "center",
          flexDirection: "column",
          border: "1px solid black",
          borderRadius: "5px",
          marginTop: "50px",
        }}
      >
        <Typography
          variant="h4"
          sx={{ fontWeight: 400, textAlign: "center", color: "#4e5de6" }}
        >
          Parking Lot
        </Typography>
        <TextField
          id="slots"
          label="Enter Car Number"
          variant="outlined"
          value={carNumber}
          fullWidth
          onChange={handleCarNumber}
          data-testid="parking-drawing-registration-input"
        />
        <Button
          variant="contained"
          onClick={handleParking}
          data-testid="parking-drawing-add-car-button"
          sx={{ marginTop: "20px" }}
        >
          Park
        </Button>
      </Container>
      <Container
        sx={{
          maxWidth: 600,
          maxHeight: 600,
          display: "flex",
          flexDirection: "column",
          flexWrap: "wrap",
          border: "1px solid black",
          borderRadius: "5px",
          marginTop: "50px",
        }}
      >
        {parkingSlots?.map((each: any) => {
          return (
            <Card
              data-testid={`parking-drawing-space-${each.id}`}
              sx={{
                width: 150,
                height: 200,
                margin: 3,
                border: "solid 1px black",
                backgroundColor: each.available ? "white" : "#5738f2",
              }}
            >
              <Typography
                variant="h6"
                textAlign="center"
                color="#edf7f7"
                fontWeight="400"
              >
                ID: {Math.floor(Math.random() * 100000)}
              </Typography>
              {!each.available ? (
                <CardContent sx={{ textAlign: "center" }}>
                  <Typography variant="body1" color="#edf7f7" fontWeight="400">
                    Spot Number: {each.id}
                  </Typography>
                  <Typography variant="body1" color="#edf7f7" fontWeight="400">
                    CarNumber:{each.carNumber}
                  </Typography>
                  <Typography variant="body1" color="#edf7f7" fontWeight="400">
                    Time:{each.time.getHours()}
                  </Typography>
                  <Button
                    variant="contained"
                    onClick={() => handleExit(each.id)}
                  >
                    Exit
                  </Button>
                </CardContent>
              ) : null}
            </Card>
          );
        })}
      </Container>
    </>
  );
};

export default ParkingLot;
