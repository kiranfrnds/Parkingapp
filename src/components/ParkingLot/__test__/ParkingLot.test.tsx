import { fireEvent, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ParkingLot from "../ParkingLot";

const mockedUsedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUsedNavigate,
}));
beforeAll(() => {
  jest.spyOn(global.console, "error").mockImplementation(() => {});
});

describe("ParkingLot", () => {
  it("should render", () => {
    render(<ParkingLot />);
    expect(screen.getByText("Parking Lot")).toBeInTheDocument();
  });
  it("should render parking-drawing-add-car-button", () => {
    render(<ParkingLot />);
    expect(
      screen.getByTestId("parking-drawing-add-car-button")
    ).toBeInTheDocument();
  });
  it("should render car number input", () => {
    render(<ParkingLot />);
    expect(screen.getByLabelText("Enter Car Number")).toBeInTheDocument();
  });
  it("should render parking-drawing-registration-input", () => {
    render(<ParkingLot />);
    expect(
      screen.getByTestId("parking-drawing-registration-input")
    ).toBeInTheDocument();
  });
  it("should alot parking slot", () => {
    render(<ParkingLot />);
    const carNumber = "KA-01-HH-1234";
    userEvent.type(screen.getByLabelText("Enter Car Number"), carNumber);
    userEvent.click(screen.getByTestId("parking-drawing-add-car-button"));
    expect(screen.getByText(carNumber)).toBeInTheDocument();
  });
});
