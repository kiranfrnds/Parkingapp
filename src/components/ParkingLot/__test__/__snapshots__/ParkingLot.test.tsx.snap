// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParkingLot should render parking slots 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-14q7dda-MuiContainer-root"
  >
    <h4
      class="MuiTypography-root MuiTypography-h4 css-1890mk3-MuiTypography-root"
    >
      Parking Lot
    </h4>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      data-testid="parking-drawing-registration-input"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="slots"
        id="slots-label"
      >
        Enter Car Number
      </label>
      <div
        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id="slots"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-1ftyaf0"
          >
            <span>
              Enter Car Number
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-1nclpby-MuiButtonBase-root-MuiButton-root"
      data-testid="parking-drawing-add-car-button"
      tabindex="0"
      type="button"
    >
      Park
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-xkho84-MuiContainer-root"
  />
</DocumentFragment>
`;
